src/LinearSolvers/ConjugateGradientSolver.jl

```diff
@@ -0,0 +1,469 @@
+- `solver`: (struct). This is a scruct for dispatch, in this case for ConjugateGradient
+- `threshold`: (float). Either an absolute or relative tolerance
+- `applyPC!`: (function). Applies a preconditioner to objecy "y" and overwrites object "z". applyPC!(z,y)
+- `args...`: (arbitrary). This is necessary for the linearoperator! function which has a signature linearoperator!(b, x, args....)
+
+# Keyword Arguments
+
+- There are no keyword arguments
+
+# Return
+- `converged`: (bool). A boolean to say whether or not the iterative solver has converged.
+- `iteration`: (int). Iteration number for the iterative solver
+- `threshold`: (float). The value of the residual for the first timestep
+
+# Comment
+- This function does conjugate gradient
+
+"""
+function LS.doiteration!(
+    linearoperator!,
+    Q,
+    Qrhs,
+    solver::ConjugateGradient,
+    threshold,
+    args...;
+    applyPC! = (x, y) -> x .= y,
+)
+
+
+    # unroll names for convenience
+
+    rtol = solver.rtol
+    atol = solver.atol
+    residual_norm = typemax(eltype(Q))
+    dims = solver.dims
+    converged = false
+
+    max_iter = solver.max_iter
+    r0 = solver.r0
+    z0 = solver.z0
+    p0 = solver.p0
+    r1 = solver.r1
+    z1 = solver.z1
+    p1 = solver.p1
+    Lp = solver.Lp
+    alias_p0 = solver.alias_p0
+    alias_Lp = solver.alias_Lp
+    alias_Q = reshape(Q, solver.reshape_tuple)
+
+    # Smack residual by linear operator
+    linearoperator!(alias_Lp, Q, args...)
+    # make sure that arrays are of the appropriate size
+    alias_p0 .= Qrhs
+    r0 .= p0 - Lp
+    # apply the preconditioner
+    applyPC!(z0, r0)
+    # update p0
+    p0 .= z0
+
+    # TODO: FIX THIS
+    absolute_residual = maximum(sqrt.(sum(r0 .* r0, dims = dims)))
+    relative_residual =
+        absolute_residual / maximum(sqrt.(sum(Qrhs .* Qrhs, dims = :)))
+    # TODO: FIX THIS
+    if (absolute_residual <= atol) || (relative_residual <= rtol)
+        # wow! what a great guess
+        converged = true
+        return converged, 1, absolute_residual
+    end
+
+    for j in 1:max_iter
+        linearoperator!(alias_Lp, alias_p0, args...)
+
+        α = sum(r0 .* z0, dims = dims) ./ sum(p0 .* Lp, dims = dims)
+
+        # Update along preconditioned direction, (note that broadcast will indeed work as expected)
+        @. alias_Q += α * p0
+
+        @. r1 = r0 - α * Lp
+
+        # TODO: FIX THIS
+        absolute_residual = maximum(sqrt.(sum(r1 .* r1, dims = dims)))
+        relative_residual =
+            absolute_residual / maximum(sqrt.(sum(Qrhs .* Qrhs, dims = :)))
+        # TODO: FIX THIS
+        converged = false
+        if (absolute_residual <= atol) || (relative_residual <= rtol)
+            converged = true
+            return converged, j, absolute_residual
+        end
+
+        applyPC!(z1, r1)
+
+        β = sum(z1 .* r1, dims = dims) ./ sum(z0 .* r0, dims = dims)
+
+        # Update
+        @. p0 = z1 + β * p0
+        @. z0 = z1
+        @. r0 = r1
```

Comment:
Have you explored the cost of not fusing the kernels in this section of the code?

Reply:
Not yet! We are planning on doing some performance checks later on in the week. We figured to get something that works into the code-base first, and then optimize later if necessary.